packages/core/src/interpreter.ts

```diff
@@ -568,30 +578,82 @@ export class Interpreter<
       return this;
     }
 
-    [...this.state.configuration]
-      .sort((a, b) => b.order - a.order)
-      .forEach((stateNode) => {
-        for (const action of stateNode.definition.exit) {
-          this.exec(action, this.state);
-        }
-      });
-
-    // Stop all children
-    this.children.forEach((child) => {
-      if (isFunction(child.stop)) {
-        child.stop();
-      }
-    });
+    this.initialized = false;
+    this.status = InterpreterStatus.Stopped;
 
-    // Cancel all delayed events
+    // we are going to stop within the current sync frame
+    // so we can safely just cancel this here as nothing async should be fired anyway
     for (const key of Object.keys(this.delayedEventsMap)) {
       this.clock.clearTimeout(this.delayedEventsMap[key]);
     }
 
+    // clear everything that might be enqueued
     this.scheduler.clear();
-    this.initialized = false;
-    this.status = InterpreterStatus.Stopped;
-    registry.free(this.sessionId);
+
+    // at the same time let what is currently processed to be finished
+    this.scheduler.schedule(() => {
+      const _event = toSCXMLEvent({ type: 'xstate.stop' }) as any;
+
+      const nextState = serviceScope.provide(this, () => {
+        const exitActions = flatten(
+          [...this.state.configuration]
+            .sort((a, b) => b.order - a.order)
+            .map((stateNode) =>
+              toActionObjects(
+                stateNode.onExit,
+                this.machine.options.actions as any
+              )
+            )
+        );
+
+        const [resolvedActions, updatedContext] = resolveActions(
+          this.machine as any,
+          this.state,
+          this.state.context,
+          _event,
+          exitActions,
+          this.machine.config.preserveActionOrder
+        );
+
+        const newState = new State<TContext, TEvent, TStateSchema, TTypestate>({
+          value: this.state.value,
```

Comment:
I think we can deviate from SCXML here, since we assume the final state to be the state _before_ exiting all states of the machine completely, and having a state of "nothingness" (empty configuration, null state value) is not very practical, especially since `state.done === true` implicitly gives us that information already.

Reply:
I've adjusted the implementation slightly and now `state.done` is equivalent to `isInFinalState(state)` - so a stopped machine doesn't end up with `state.done === true`